### ISSUE TITLE
Strange MathJax rendering when using tablet (Samsung Galaxy Tab S6)

### ISSUE THREAD

**Issue by hugomarins** [ankidroid/Anki-Android#7848]
###### Reproduction Steps

1. Reviewing cards with MathJax.

###### Expected Result
MathJax being rendered regularly, like it is in my phone Samsung Note 9.

###### Actual Result
When reviewing in the tablet (Samsung Galaxy Tab S6), MathJax is rendered strangely:
![SmartSelect_20201208-054928_AnkiDroid](https://user-images.githubusercontent.com/54500828/101624308-89b5d580-39f8-11eb-9da0-4145a73e17e7.jpg)
![SmartSelect_20201208-061155_AnkiDroid](https://user-images.githubusercontent.com/54500828/101624311-8a4e6c00-39f8-11eb-8f34-819f449932fe.jpg)
![SmartSelect_20201208-083607_AnkiDroid](https://user-images.githubusercontent.com/54500828/101624313-8ae70280-39f8-11eb-8d6f-ae1a03350af2.jpg)

###### Debug info
AnkiDroid Version = 2.15alpha11
Android Version = 10
ACRA UUID = 38f97252-6295-4be3-b0f4-864eaa211b02
Scheduler = std2
Crash Reports Enabled = true

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by mikehardy** [comment]
Very interesting - are they both on the alpha? Mathjax just had a massive change merged: #7782 

Following that, if they are on different versions, then it would be interesting to see if 2.14.2 vs 2.15alpha is the difference by installing the opposite of whatever is currently on each device (note this is easy and user-comfortable, we publish "parallel" builds so you can just install the '.A' of whatever version you want to try and it runs side-by-side with the main version, against the same collection - https://github.com/ankidroid/Anki-Android/releases

If they are on same version and it's a form factor issue then that is most strange. I wonder if AnkiWeb has the same issue when accessed via the mobile device?

Basically just looking to isolate exactly where the problem is as a rough cut

**Comment by hugomarins** [comment]
Mike, both of them in the latest alpha version. I've been using the tablet (Samsung Galaxy Tab S6) since Nov 19th, 2020, always in the alpha version, and since that day this situation happens.

I tried testing a parallel version (2.14.2) in the tablet, but it didn't change the situation.

By "I wonder if AnkiWeb has the same issue when accessed via the mobile device?", did you mean accessing AnkiWeb via browser? I never use it, because it do not generate MathJax for me, nor in the mobile nor in the computer. It looks like:

![image](https://user-images.githubusercontent.com/54500828/101674760-aa515000-3a37-11eb-872b-f19c497bc455.png)
![AnkiWeb Mathjax](https://user-images.githubusercontent.com/54500828/101675099-1e8bf380-3a38-11eb-91b8-94f78dd9eba5.jpg)

Thanks!

**Comment by mikehardy** [comment]
Interesting, did not know that AnkiWeb didn't handle Mathjax, that's a bummer.

At any rate, thank you very much for testing 2.14.x that tells me that we likely have some sort of thing under our control with respect to display sizes, vs some unknown Mathjax library issue between v2 and v3 of Mathjax. That's valuable knowledge.

Not sure when someone will have time to look at it, but my first thing to do would be to create a tablet emulator via Android Studio AVD manager and have it up next to a phone form factor emulator to reproduce things, then I'd go from there...

**Comment by Arthur-Milchior** [comment]
In case you didn't already do it, I suggest you putting a message about ankiweb and mathjax on anki forum. I suspect that adding mathjax to ankiweb would be really simple to do and so hopefully, we could get it quickly

Would you mind exporting a card with the problem, so I can take a look myself. You can export a single card from the computer, by selecting the card in the browser.

@hgiesel, it seems you know a little bit about MathJax, would you have any idea what's going on?

**Comment by hgiesel** [comment]
This seems to be an issue with the web view on the Samsumg Galaxy Tab. 
There's two output formats for MathJax right now: `chtml` (Common HTML), which is basically HTML with very elaborate CSS, and `svg`, which is dynamically rendered SVG. All anki platforms (even iOS which is still on MathJax 2, as of writing this) use chtml because it is much faster.

However chtml needs to be very precisely calibrated. For example, it seems like the root symbol `√` would be one component, but actually it is Unicode character + a border on the top of the rest:

![Screenshot 2020-12-15 at 14 17 25](https://user-images.githubusercontent.com/7188844/102219806-423fb580-3ee0-11eb-8147-5e4940330c7f.png)

Just by messing with the paddings and margins a little bit (like changing a padding of `0.758` to `0.958`, etc.), I can displace all elements a little, making it seem off:

![Screenshot 2020-12-15 at 14 27 27](https://user-images.githubusercontent.com/7188844/102220854-a9aa3500-3ee1-11eb-9a5e-bb919d059d95.png)

So to summarize: I think MathJax does not exactly support your web environment, and so makes an educated guess.
You could try out this [demo website with CHTML rendering](https://mathjax.github.io/MathJax-demos-web/tex-chtml.html) vs this [demo website using SVG rendering](https://mathjax.github.io/MathJax-demos-web/tex-svg.html) on Galaxy Tab, but if possible with the built-in browser (which I would assume is a Samsung browser?)

**Comment by hugomarins** [comment]
@Arthur-Milchior Ok, I'll post the issue about MathJax in Ankiweb forum.

Here's some cards that render strangely:
[Selected Notes MathJax rendering.zip](https://github.com/ankidroid/Anki-Android/files/5696558/Selected.Notes.MathJax.rendering.zip)

And how they look like in my Samsung Tab S6:
![1](https://user-images.githubusercontent.com/54500828/102229521-a9526f80-3eca-11eb-996e-bbbb8e4d9ab1.jpg)
![2](https://user-images.githubusercontent.com/54500828/102229527-ab1c3300-3eca-11eb-961b-6f7bf0ee72d8.jpg)
![3](https://user-images.githubusercontent.com/54500828/102229531-abb4c980-3eca-11eb-9099-903f0376cc81.jpg)
![4](https://user-images.githubusercontent.com/54500828/102229533-ac4d6000-3eca-11eb-96c5-be6ba6d59c19.jpg)

@hgiesel Thanks for your help. I tried out the demo website of both chtml and svg rendering, in both Samsung buit-in browser and Chrome in my samsung Tab S6, and they looked nice. Only in Ankidroid it goes a little strange (especially square roots):
![11](https://user-images.githubusercontent.com/54500828/102230514-bb80dd80-3ecb-11eb-98fd-1e1445736783.jpg)
![12](https://user-images.githubusercontent.com/54500828/102230520-bcb20a80-3ecb-11eb-97b5-611256bab6e4.jpg)
![13](https://user-images.githubusercontent.com/54500828/102230522-bd4aa100-3ecb-11eb-9767-25b99eef2059.jpg)
![14](https://user-images.githubusercontent.com/54500828/102230527-bde33780-3ecb-11eb-933e-8ffce865841f.jpg)

**Comment by Arthur-Milchior** [comment]
I was able to reproduce on SM-T515(Tab A 10.1) but not on SM-A202F(Samsung A20e)

Beautiful card by the way. May I ask whether it's a shared deck ?

**Comment by hugomarins** [comment]
@Arthur-Milchior this is not a shared deck. These note types and card templates I developed throughout the time, sometimes based in css of other people, adapting to my needs. I'm applying to get a given position, and in Brazil that is possible only by succeeding in a formal test. That's what I'm preparing for. And so I adapted Anki to help me in this journey. Thanks for helping!

**Comment by Arthur-Milchior** [comment]
I posted a bug report in https://github.com/mathjax/MathJax/issues/2583
I find hard to recreate a minimal bug, since it occurs in webview and not in the browser.

To be clear, @hugomarins , I've no idea how much time it'll take to debug this one. I hope somebody else will be able to come and tell us it's easy to fix. We won't ship a beta with this problem, in the worse case I assume we'll reverse to previous, slower but more correct, math rendering. However, for the alpha, I've no idea what we'll do.

Can you please let us know how big of a problem it is. Do you feel like you can still use the app, or that you'll need to go back to a stable version because it's not usable anymore?

**Comment by hugomarins** [comment]
@Arthur-Milchior Yes, definitely I can still use the app. It just bother a little, as it is not visually ok, but I can still perfectly read the formulas. I just wanted to inform the occurrence of the matter to you guys. Thanks!

**Comment by hgiesel** [comment]
@Arthur-Milchior I'm pretty sure @hugomarins mentioned that the issue is caused by both MathJax 2, and MathJax 3:

> I tried testing a parallel version (2.14.2) in the tablet, but it didn't change the situation.

So reverting to MathJax 2 is not the solution here. One obvious solution, even though it might be overkill, is supporting rendering to svg, which would certainly not have these display issues, but might introduce its own [slew of issues](http://docs.mathjax.org/en/latest/output/svg.html).

**Comment by Arthur-Milchior** [comment]
Oh, you are perfectly right. I didn't realize that 14.2 means Mathjax 2. My bad.

I'm surprised, I just tried Hugo's collection with 14.2 and I can't reproduce. @hugomarins , can you please confirm that you already had the same problem with version 14.2


I can't reproduce the problem with SVG.
The issue in your link are:
* not compatible with IE: (we dno't care)
* change in tag number. Cards are supposed to be small, I doubt that this feature is used a lot, and a change does not seem so bad
* no copyable. That's troublesome, but I would not expect it to be a core feature.  

So I'd be fine with those tradeoff. Assuming I didn't miss any other.

**Comment by mikehardy** [comment]
-1 on a revert, I don't think that will solve it no, and I have a strong bias against going backwards unless absolutely necessary

+1 on maybe an SVG output toggle if it doesn't seem too hard to do. I read that link of possible issues (thank you Henrik!) and it didn't seem like they would have much impact in our context (resizing? IE9 issues?) so it may be the right format to use by default even to avoid the font issues that were mentioned. At the very least allowing it as a toggle in advanced prefs could work (assuming - again - it doesn't seem too hard to do)

If it seems hard to do SVG format this may be a rough edge we just have to live with for now

**Comment by Arthur-Milchior** [comment]
SVG is not hard. As I wrote, I can't reproduce in SVG, which implicitly meant that I tried with SVG and did it.

I never realized that we can't copy images from the webview. I wonder why. (Independently of whether it's svg or not)

I'll try to work with mathjax team if they are interested, because it's still a bug that would be interesting to solve.

**Comment by hugomarins** [comment]
@Arthur-Milchior Sorry to take this long to answer your question. Yes, I noted the problem in 14.2 also. Like below:
![Screenshot_20201220-070852_AnkiDroidA](https://user-images.githubusercontent.com/54500828/102710855-1a5fa180-4294-11eb-97cf-33078c3fba87.jpg)
![Screenshot_20201220-070948_AnkiDroidA](https://user-images.githubusercontent.com/54500828/102710858-1d5a9200-4294-11eb-8f6d-0e4803605c82.jpg)
![Screenshot_20201220-071036_AnkiDroidA](https://user-images.githubusercontent.com/54500828/102710864-277c9080-4294-11eb-9c59-f5b7f805b53d.jpg)

I saw you released a new version to fix it. I'll try and let you know. Thanks!

**Comment by Arthur-Milchior** [comment]
> @Arthur-Milchior Sorry to take this long to answer your question.

No problem. Since it does not seems to cause any data loss, any leak of personal information, and it's in alpha, there is absolutely no emergency here. You can takes as much time as you want.

Actually, if the bug is not even new, then there is even less emergency as it won't block an eventual version 15 realease I suspect.


>  Yes, I noted the problem in 14.2 also. 

Thank you very much for the screenshots. That's really appreciated. I'll try to reproduce on my tablet with this card later. I'm really surprised, I saw a lot of maths and my tablet and never noticed anything like that.

> I saw you released a new version to fix it. I'll try and let you know. Thanks!

Sadly, I can already tell you that there are bugs reported with this new version. Hopefully, it'll still work for you and that would be interesting that you confirm it.
